admin/python/lsst/qserv/testing/config.py

```diff
@@ -0,0 +1,299 @@
+    def __init__(self, path, mode="random"):
+        # read all numbers from file as integers
+        if path == "/dev/null":
+            # for testing only
+            self._array = [0]
+        else:
+            self._array = np.fromfile(path, dtype=int, sep=" ")
+        self._mode = mode
+        assert mode in ("random", "sequential")
+        self._seq = 0
+
+    def __call__(self):
+        if self._mode == "random":
+            return random.choice(self._array)
+        else:
+            if self._seq >= len(self._array):
+                self._seq = 0
+            value = self._array[self._seq]
+            self._seq += 1
+            return value
+
+
+class QueryFactory:
+    """Class which generates queries.
+
+    Parameters
+    ----------
+    txt : `str`
+        If ``variables`` is ``None`` then full query text, otherwise it is a
+        query template.
+    variables : `dict` [`str`, `dict`]
+        Dictionary whose keys are variable names and values are dictionaries
+        with a description of how to generate variable value.
+    """
+    def __init__(self, txt, variables=None):
+        self._txt = txt
+        self._vars = {}
+        if variables is not None:
+
+            for var, config in variables.items():
+                generator = None
+                if "distribution" in config:
+                    if config["distribution"] == "uniform":
+                        min = config.get("min", 0.)
+                        max = config.get("max", 1.)
+                        generator = _ValueRandomUniform(min, max)
+                    elif config["distribution"] == "uniform_int":
+                        min = config.get("min", 0)
+                        max = config.get("max", 2**64)
+                        generator = _ValueRandomUniformInt(min, max)
+                elif "path" in config:
+                    path = config["path"]
+                    mode = config.get("mode", "random")
+                    generator = _ValueIntFromFile(path, mode)
+                if generator is None:
+                    raise ValueError(f"Cannot parse variable configuration {var} = {config}")
+                self._vars[var] = generator
+
+    def query(self):
+        if not self._vars:
+            return self._txt
+        else:
+            values = {}
+            for var, generator in self._vars.items():
+                values[var] = generator()
+            return self._txt.format(**values)
+
+
+class Config:
+    """Configuration of test harness.
+
+    Parameters
+    ----------
+    configs : `list` [`dict`]
+        List of dictionaries, cannot be empty.
+    """
+    def __init__(self, configs):
+
+        if not configs:
+            raise ValueError("empty configurations list")
+
+        # merge them
+        self._config = {}
+        for config in configs:
+            self._config = self._merge(self._config, config)
+
+        classes = set(self._config.get("queryClasses", {}).keys())
```

Comment:
Here I'm personally right at the edge of my tolerance for nested containers and functional manipulation of them. IMO it may be time now or soon to consider writing a proper config class.

Reply:
...well, the api to this class is good so 🤷.

Reply:
"The proper config class" is this class itself. We need to read data from YAML in any case and this is what this class does. Clients use API which does not expose this detail.